### PR TITLE
placement: api for adding rules in one request

### PR DESCRIPTION
**Pull request by xhebox**
### What problem does this PR solve?

Related to #2680.

Now one can post an array of rules to `/config/rules`, to set all rules in one request. In case there is an error, all modifications are cancelled.

### Check List

Tests

- Unit test
- Integration test

Code changes

- Has HTTP API interfaces change

### Release note

- atomic POST api `/config/rules` for setting all rules in one request, parameter is an array of rules

### CODE REVIEW

**Comment by disksing** [comment]
Thanks, @xhebox , very high-quality code. Not only correct but also elegant.

File: server/schedule/placement/rule_manager.go [hunk 1]
```diff
@@ -273,3 +274,53 @@ func (m *RuleManager) FitRegion(stores StoreSet, region *core.RegionInfo) *Regio
...
+		}
+	}
+
+	m.Lock()
+	defer m.Unlock()
+
+	oldRules := make(map[[2]string]*Rule)
+
+	for _, rule := range rules {
+		oldRules[rule.Key()] = m.swapRule(rule)
```
Comment (nolouch) [inline server/schedule/placement/rule_manager.go:299]: May `oldRules` be nils? will it build a `rulelist` successfully if `oldRules` contains nil pointer? we need to add a test for it.

Comment (xhebox) [reply server/schedule/placement/rule_manager.go:299]: Yes. The first case, actually, should already have `oldRules` full of nils. But its ID is duplicated with `SetRule`.

**Pull request merged**